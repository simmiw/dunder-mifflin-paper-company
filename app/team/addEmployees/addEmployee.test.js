import Form from "./addEmployee";
import { render, fireEvent, waitFor, screen } from "@testing-library/react";

const mockRouterPushFn = jest.fn();

jest.mock("next/navigation", () => ({
  useRouter: () => ({
    push: mockRouterPushFn,
    refresh: jest.fn(),
  }),
}));

describe("addEmployee", () => {
  let sut;

  beforeEach(() => sut = render(<Form />));

  it("should render the form", () => {
    expect(screen.getByLabelText(/Name/)).toBeVisible();
    expect(screen.getByLabelText(/Email/)).toBeVisible();
    expect(screen.getByLabelText(/Title/)).toBeVisible();
    expect(screen.getByLabelText(/Branch/)).toBeVisible();
    expect(screen.getByLabelText(/Team/)).toBeVisible();
    expect(screen.getByLabelText(/Image/)).toBeVisible();
    expect(screen.getByLabelText(/Know More/)).toBeVisible();
    expect(screen.getByText(/Submit/)).toBeVisible();
    expect(sut.container).toMatchSnapshot();
  });


  it("should submit the form when filled and valid", async () => {
    global.fetch = jest.fn().mockResolvedValue({ status: 201 });

    fireEvent.change(screen.getByLabelText(/Name/), {
      target: { value: "Michel Scott" },
    });

    fireEvent.change(screen.getByLabelText(/Email/), {
      target: { value: "michalscoot@gmail.com" },
    });

    fireEvent.change(screen.getByLabelText(/Title/), {
      target: { value: "Regional Manager" },
    });

    fireEvent.change(screen.getByLabelText(/Branch/), {
      target: { value: "Scanton" },
    });

    fireEvent.change(screen.getByLabelText(/Team/), {
      target: { value: "Sales" },
    });

    fireEvent.change(screen.getByLabelText(/Image/), {
      target: { value: "image.jpg" },
    });

    fireEvent.change(screen.getByLabelText(/Know More/), {
      target: { value: "More Details" },
    });

    fireEvent.click(screen.getByText(/Submit/));

    expect(global.fetch).toHaveBeenCalledWith(
      "http://localhost:4000/employees",
      {
        method: "POST",
        headers: { "content-type": "application/json" },
        body: JSON.stringify({
          name: "Michel Scott",
          email: "michalscoot@gmail.com",
          title: "Regional Manager",
          branch: "Scanton",
          department: "Sales",
          thumbnail: "image.jpg",
          about: "More Details",
        }),
      }
    );
    expect(screen.getByText("Adding...")).toBeVisible();
    await waitFor(() => {
      expect(mockRouterPushFn).toBeCalledWith("/team");
    });
  });
});
