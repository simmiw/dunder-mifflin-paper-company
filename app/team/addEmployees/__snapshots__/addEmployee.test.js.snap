// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addEmployee should render the form 1`] = `
<div>
  <main>
    <form
      class="w-1/2"
    >
      <label>
        <span>
          Name:
        </span>
        <input
          required=""
          type="text"
          value=""
        />
      </label>
      <label>
        <span>
          Email:
        </span>
        <input
          required=""
          type="text"
          value=""
        />
      </label>
      <label>
        <span>
          Title:
        </span>
        <input
          required=""
          type="text"
          value=""
        />
      </label>
      <label>
        <span>
          Branch:
        </span>
        <input
          required=""
          type="text"
          value=""
        />
      </label>
      <label>
        <span>
          Team:
        </span>
        <input
          required=""
          type="text"
          value=""
        />
      </label>
      <label>
        <span>
          Image:
        </span>
        <input
          required=""
          type="text"
          value=""
        />
      </label>
      <label>
        <span>
          Know More:
        </span>
        <input
          required=""
          type="text"
          value=""
        />
      </label>
      <button
        class="btn-primary"
      >
        <span>
          Submit
        </span>
      </button>
    </form>
  </main>
</div>
`;
